content/post/reliable-change-detection.md

```diff
@@ -0,0 +1,258 @@
+Tools like this must ultimately be benchmarked under *load*, rather than merely measured in terms of naive iteration speed.
+
+The "reliable" criteria (5, 6 and 7) deserve a bit more attention as well. Change detection must work:
+
+1. If the change occurred immediately before the detecting system.
+2. If the change occurred later in the frame than the detecting system. This results in a frame delay, but shouldn't cause catastrophic failure.
+3. If the detecting system runs more than once each frame. This can occur with looping run criteria, commonly in `States`.
+4. If the detecting system didn't run for one or more frames.
+
+The final criteria is both surprisingly devious and important.
+Skipping systems via **run criteria** is a powerful tool for common game functionality, like pausing, swapping between modes using `States`, or having systems that only run every X seconds or frames.
+It would be very nice if these systems could take advantage of our change detection as well.
+However, that introduces another serious complication: changes could be "seen" by some systems and marked as complete well before they're seen by another system that was skipped.
+
+With these constraints in mind, let's take a look at some attempts at a solution.
+
+## Rejected designs
+
+As you may have guessed from the constraint list, this problem is trickier than it first appears.
+Let's start with some very simple solutions, and slowly show why they're inadequate.
+
+### Solution 0: Treat everything as changed
+
+We simply throw up our hands and write off the feature.
+This is extremely ergonomic, incredibly high performance, and never results in a false negative!
+Of course, it fails the false positive criteria completely and must be written off.
+
+### Solution 1: Changes persist until the end of the frame
+
+This was the solution as of Bevy 0.4.
+It passes criteria 1-4, as it is ergonomic and performant.
+It even passes criteria 7, as changes are immediately marked.
+
+However, as discussed above, this solution fails to be reliable.
+Unsurprisingly, it fails criteria 5: false negatives can exist and are quite common, occurring whenever the change detecting system occurs earlier in the frame than the changes are made.
+False negatives also occur when systems are skipped for one or more frame.
+
+More surprisingly, it also fails criteria 6, presenting false positives!
+This occurs whenever a system is run more than once per frame: it will process a change on the first pass, and then find that the same component is still marked as changed on all future passes.
+This can have subtle but catastrophic effects if your systems that respond to change are not [idempotent](https://stackoverflow.com/questions/1077412/what-is-an-idempotent-operation), causing changes to be double-counted and overcompensated.
+
+### Solution 2: Changes persist until the end of the frame after they are made, but you can only detect changes in the previous frame
```

Comment:
Quite a long heading, does this cause layout problems on the website? Might be fine though